### PR TITLE
Translatable with fallbacks

### PR DESCRIPTION
**Pull request by pxpm**
### BEFORE - What was wrong? What was happening before this PR?

replaces #4262 

Applies the solution me and @tabacitu discussed detailed here: https://github.com/Laravel-Backpack/CRUD/issues/4100#issuecomment-1067671985

### CODE REVIEW

**Comment by tabacitu** [comment]
Thank you @pxpm . I'll take a look as soon as we're done with the MUSTs in the v5 board. Please take a look at the email I just sent for perspective 💪 Let's do this!!!

**Comment by kronos-collector** [comment]
Is there any way this fix be extended to the show.blade.php as well? It corrects the functionality on edit but you can still only preview/show the default locale.

**Comment by tabacitu** [comment]
I like this solution. I think it's much better than what we had before. It's arguable if it's ok to introduce this as a non-breaking change... but I think if we are _smart_ about it, and the interface is clear enough... it's ok. Admins will understand that the default behaviour has changed, and the admin panel no longer translates the entity by default, but that's still available with a click.

To make it 100% clear and intuitive, we need to be super-careful about the interface. That's where I would like us to make quite a few improvements @pxpm . I want us to go from this:
CleanShot 2022-08-03 at 09 40 29

To this:

CleanShot 2022-08-03 at 09 40 47

Note the most important differences:
- put it in an alert, on the same line as the Language dropdown, with soft padding and nicely ordered;
- action is a button, not a link; that makes it more "clickable" and gives it higher visual hierarchy; same color means it's related to the button to the right (language picker);
- the button text should say "_Fill inputs from {fallback_language}_" not "_Load defaults_"; that's because that's what it does, it just fills the inputs; by "_Load default_" people could understand the translation is already saved... which is not true; plus "defaults" is ambiguous;
- the text should also note the current language, so that it's clear what translation we're talking about;

**Comment by jorgetwgroup** [comment]
Hi @pxpm i need to be honest, i never use early translate feature.

I try now as you describe the problem, and yes is annoying and strange:

Create english:

screenshot-backpack test-2022 09 12-10_44_13

Translate french and i deleted "description", but when i came back to edit, still there, in DB is null:

screenshot-backpack test-2022 09 12-10_44_51

I update to branch "translatable-with-fallbacks" and works as normal:

screenshot-backpack test-2022 09 12-10_45_47

I create new entry and everything works great!

screenshot-backpack test-2022 09 12-10_46_50

screenshot-backpack test-2022 09 12-10_48_24

Look ready for me.

Cheers.

**Comment by tabacitu** [comment]
@pxpm please do a last test on this one (it was built 3 months ago). I you're ok with it, MERGE IT!

git-merge-cfkifmdg3ifoi

**Comment by pxpm** [comment]
We should make sure, that when you add the translation to an already existing untranslated project, we assume that the current strings in database make for the default value in default language when the json is not built yet. 

Comes from https://github.com/Laravel-Backpack/community-forum/issues/401

File: src/resources/views/crud/edit.blade.php
Comment (tabacitu) [inline src/resources/views/crud/edit.blade.php:113]: A lot of code was added here. Perhaps we should move it to a different file? That would make the change smaller AND provide a way to edit that without changing the entire edit view.

I'd say `resources/views/edit_translation_notice.blade.php` or something like that.

File: src/config/backpack/operations/list.php
Comment (tabacitu) [inline src/config/backpack/operations/list.php:71]: Love that it's configurable. Though... I don't know if it's necessary at all, tbh. 

If EN is the fallback/default, I would NEVER expect the list/show operation to show "english-version" for Romanian/Portuguese. The way I see it, that is only something we did on the UpdateOperation to make it easier to translate. It would become a shit-show if we did that on List/Show... 

Why? Because afaik, when you use the model inside the app itself... it won't use the fallback locale. Right? 👀 If you do `$product->title` and that's not translated to PT... it will show... empty?! I'm asking!

File: src/app/Http/Controllers/Operations/ListOperation.php
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/ListOperation.php:138]: Wait. I don't understand. Why does this PR make changes to the ListOperation? The bug was about changing the UpdateOperation behaviour... not the ListOperation... right?

File: src/app/Http/Controllers/Operations/ShowOperation.php
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/ShowOperation.php:84]: Same here. Why does this PR make changes to the ShowOperation?

File: src/app/Library/CrudPanel/Traits/Search.php
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Search.php:393]: I spy with my little eye... this PR making changes in other operations than Update.

**Review by tabacitu — COMMENTED** [review]
Appologies for taking a long time to review this Pedro. But I gave it a look on Tuesday and my spidey-sense was triggered. Wanted to sleep on it. After I looked at it today, I think I know what triggered my spidey-sense, let's see if my concerns are valid:

Seems like this makes changes in other operations that UpdateOperation. Namely List and Show. Why? Is that really necessary? From [my last comment on it](https://github.com/Laravel-Backpack/CRUD/issues/4100#issuecomment-1067671985) I was under the impression we only needed to change the behaviour on the UpdateOperation, for translatable items. Which would be like 10% of users. The way this is coded now, it would affect 100% of all users. Some translatable methods are called even if they're not using translatable... for all major operations. Yikes! See my other inline comment where I ask if that's necessary to change ListOperation and ShowOperation, let's investigate a bit. Maybe I have my green light and don't remember it 🤷‍♂️

File: src/app/Http/Controllers/Operations/ListOperation.php
Comment (pxpm) [reply src/app/Http/Controllers/Operations/ListOperation.php:138]: My understanding the bug was in the fact that we didn't set locales on models. (in any operation).
In the update operation it's not enough set the locale on model, so we also provided a way to fill the inputs from other locales. 
In all other operations if you see, I am just setting the locale on model, and created that optional setting I didn't touch anything else in those operations.

But if you feel it don't belong here as I think it does, I can separate those changes into a different PR.

File: src/app/Http/Controllers/Operations/ShowOperation.php
Comment (pxpm) [reply src/app/Http/Controllers/Operations/ShowOperation.php:84]: The exact same reason I explained above ... but your call 👍

File: src/config/backpack/operations/list.php
Comment (tabacitu) [reply src/config/backpack/operations/list.php:71]: Oh damn! Excellent example! 👏👏👏 1000 words 

No, I do not expect to see empty stuff... but I don't quite expect to see them in the fallback language either... not without a notice or something that makes me understand "_RO version not available - here's the EN one_". But I don't like that option very much either, would further complicate how this works. Having a prefix like `RO missing. EN: _` would probably look like crap, right? 😔

Hmm... ok... I guess `useFallbackLocale()` is reasonable way to approach this then. Good find!

---

One more question so I understand this scenario completely... I (Cristian) have my admin panel in RO, I see the Articles translated in the EN (the fallback locale). I click Edit... what do I see? I expect Article with empty inputs right?

Mind shooting a quick video/loom/gif for me with the whole process, so I understand?

**Comment by tabacitu** [comment]
> My understanding the bug was in the fact that we didn't set locales on models. (in any operation).

I've just gone through the bug report. To me it does NOT look like anybody has requested a `fallbackLocale`. [The reported problem](https://github.com/Laravel-Backpack/CRUD/issues/4100) was that we pre-populated the fields with the fallback.

I don't remember planning/agreeing to create this `useFallbackLocale` that spiders into every operation... but then again you know me, even if we did talk about it, I wouldn't remember it 😅

I don't think having `useFallbackLocale = true` is a good idea. I think it makes the whole translatable experience a lot more confusing. I wouldn't recommend anybody use `useFallbackLocale`, and I would not agree to make it the default in the next version either. So then the question becomes... why do we have it in the first place?

1) Do you remember @pxpm ? How did this idea come up? Why do we have this thing about using the fallback locale at the admin-panel/UI level... when Spatie has already provided a way to do this... at the model level? Afaik if Spatie has provided this at the model-level, if people DO want this fallback behaviour, but ONLY in their admin panel... we can instruct them to create a `AdminableProduct extends Product` model for their CRUD, and they do the Spatie config in that model. That would achieve the same result in that _very unlikely_ (in my opinion) scenario where the people want this, without us having to do a lot of code changes, to all our operations.

2) Right now it looks to me like it'd be better to just remove `fallbackLocale` altogether, with all its tentacles, and keep it simple by doing what we promised to do in https://github.com/Laravel-Backpack/CRUD/issues/4100#issuecomment-1067671985 . I don't have any problems with what we said we'd do there (except some beautifying). Looks like all the triggers from my spidey sense come from `useFallbackLocale`. 

Help me figure out (1) please @pxpm . Let me know if you still this is worth doing after my objections above. And how this config came to be, I really don't remember.

**Comment by pxpm** [comment]
> So it looks like this PR does a lot more than what was requested.

For me this PR streamlines the whole translation process. Every time we get a model, we set the appropriate locale on it, I really don't care if we called it show, update or list operation. We get a model (In CRUD), and we set the locale on that model. 

> I don't remember planning/agreeing to create this useFallbackLocale that spiders into every operation...
You didn't, this was my judgment call. 

> when Spatie has already provided a way to do this... at the model level? Afaik if Spatie has provided this at the model-level, if people DO want this fallback behaviour, but ONLY in their admin panel... we can instruct them to create a AdminableProduct extends Product model for their CRUD

I've proposed this solution on other situations and you never agreed to that solution. and IMO you were right and that's why we had other solution in place to don't recommend this. 

Basically our `useFallback` is an "admin configuration" for the `fallbackAny` of spatie, without the need for the developer to do any additional configuration in the model.
That's what we do in: https://github.com/Laravel-Backpack/CRUD/pull/4267/files#diff-53590d152d70c4232c2ab7c2c670a7978e146cd6179c7e407854ebe6b224fe27

Plus, the settings Spatie provides, are either: 
- Configure in the model the fallback local (for that you need to make sure your fallback locale has the translations). 
- Configure the `fallbackAny` in a service provider for the whole translation process (not per model). 

We use the later, so we have 3 layers of translation "per model". 
- Locale is translated we show it.
- FallbackLocale is translated we show it. 
- Locale and Fallback are not translated we use the fallbackAny. (behind a setting). 

I think there is no sense arguing this here. 
I will remove all that from here and leave only the update stuff, hopefully we can merge this this week. 


################################ 

Regarding what we can actually change in this PR:
> probably a good idea to make it info / blue? would look good on all themes;
Nop, it does not look good on all themes, you can try it just by changing the css classes. 

> probably a good idea to turn the button inside the bubble into a btn-link or something (still with a dropdown); it's confusing to have two buttons to change the language, one next to the other (one inside the bubble, one outside it);

I personally don't link the link, would have preferred the buttons, but 🤷

File: src/resources/views/crud/inc/edit_translation_notice.blade.php
Comment (tabacitu) [inline src/resources/views/crud/inc/edit_translation_notice.blade.php:1]: This alert needs a little more work in terms of looks. The current design looks unpolished:
CleanShot 2024-06-18 at 12 16 32@2x

I played around with it a bit, and got it to this, which I think is a little better, especially if instead of having a dropdown... we just list all the available languages, so the admin clicks on the language they want as fallback:
CleanShot 2024-06-18 at 12 15 40@2x

And then I thought... **why show this notice in a persistent bubble there... or a text that we want to standa out... when we have notification bubbles?** @pxpm what do you think if... instead of showing this on-page... we use JS to trigger a notification bubble that stays onscreen (if not dismissed), with this text? To me it sounds like:
- it would be consistent across themes, with zero effort
- it would stand out more than it does now (as a yellow warning / blue info... not sure which is better)
- it would save us from designing an element here

If you agree a notification bubble is the best solution here, @pxpm , please implement that. Otherwise, let's go with a simpler solution like my screenshot above (yes, the translation button would be to the left, even when the notice isn't there).

**Review by tabacitu — CHANGES_REQUESTED** [review]
Yes, I'm happy with the implementation now. Thank you Pedro. Let's do a first round of polish please! I see quite a few things here that could use a second look.

File: src/app/Models/Traits/SpatieTranslatable/HasTranslations.php
Comment (pxpm) [reply src/app/Models/Traits/SpatieTranslatable/HasTranslations.php]: 1 - Can't change the name as that's the name Spatie uses, see: https://github.com/spatie/laravel-translatable/blob/2b81c88beb7ca6616808d9a1cc98a0b3d0377e1a/src/HasTranslations.php#L23

2 - Done.

3 - Because Spatie defaults to true when property is not set. And yes, we use fallback locales by default (before this PR), that's why if you open an entry in any locale that's not translated you get the translation from the fallback locale instead of the empty inputs as in this PR.

**Comment by pxpm** [comment]
This is the current implementation: 
image

Added a setting that allow developer to disable this behavior called: `showTranslationNotice` that defaults to `true`. 

Please evaluate and let me know. 

Cheers

File: src/app/Models/Traits/SpatieTranslatable/HasTranslations.php
Comment (tabacitu) [reply src/app/Models/Traits/SpatieTranslatable/HasTranslations.php]: 1. ok
2. ok
3. ok - if we only use it in one place, let's do that

File: src/config/backpack/operations/update.php
Comment (tabacitu) [inline src/config/backpack/operations/update.php:40]: ❤️

File: src/config/backpack/operations/update.php
Comment (tabacitu) [inline src/config/backpack/operations/update.php:39]:
```suggestion
    // that allows the user to fill the form from another language?
```

**Review by tabacitu — APPROVED** [review]
Gave it another look. Love it. Good to go imho!

**Pull request merged**
